probe-rs/src/probe/jlink/mod.rs

```diff
@@ -335,36 +340,117 @@ impl DebugProbe for JLink {
+                    // Send the reset sequence (> 50 0-bits).    
+                    iter::repeat(true).take(64)
+                    // Send the JTAG to SWD sequence.
+                    .chain(jtag_to_swd_sequence.iter().copied())
+                    // Send the reset sequence again in case we were in SWD mode already (> 50 0-bits).
+                    .chain(iter::repeat(true).take(64))
+                    // Send 10 idle line bits.
+                    .chain(iter::repeat(false).take(10));
+
+                // Construct the direction sequence for reset sequence.
+                let direction =
+                    // Send the reset sequence (> 50 0-bits).    
+                    iter::repeat(true).take(64)
+                    // Send the JTAG to SWD sequence.
+                    .chain(iter::repeat(true).take(16))
+                    // Send the reset sequence again in case we were in SWD mode already (> 50 0-bits).
+                    .chain(iter::repeat(true).take(64))
+                    // Send 10 idle line bits.
+                    .chain(iter::repeat(true).take(10));
+
+                // Send the init sequence.
+                // We don't actually care about the response here.
+                // A read on the DPIDR will finalize the init procedure and tell us if it worked.
+                jlink.swd_io(direction, swd_io_sequence)?;
+
+                // So we read the DPIDR to complete the init sequence.
+                let response = DAPAccess::read_register(self, PortType::DebugPort, 0x0000)?;
+
+                // If we get to here, the init worked.
+                let dpidr = DPIDR(response);
+                log::debug!("Successfully read DPIDR. SWD init successful.");
+                log::debug!(
+                    "Manufacturer is {} with part_no={} and revision={}",
+                    jep106::JEP106Code::new(dpidr.jep_cc(), dpidr.jep_id())
+                        .get()
+                        .unwrap_or("Unknown JEP106 code."),
+                    dpidr.part_no(),
+                    dpidr.revision(),
+                );
+
+                // Clear all error flags in the abort register.
+                let mut abort = Abort::default();
+                abort.set_orunerrclr(true);
+                abort.set_wderrclr(true);
+                DAPAccess::write_register(
+                    self,
+                    PortType::DebugPort,
+                    Abort::ADDRESS as u16,
+                    abort.into(),
+                )?;
+
+                // Enable & power up all debug facilities.
+                let mut ctrl = Ctrl::default();
+                ctrl.set_cdbgpwrupack(true);
+                ctrl.set_cdbgpwrupreq(true);
+                DAPAccess::write_register(
+                    self,
+                    PortType::DebugPort,
+                    Ctrl::ADDRESS as u16,
+                    ctrl.into(),
+                )?;
+
+                // We are ready to debug.
```

Comment:
This is very similar to the code for the ST-Link and DAPlink probes:
 https://github.com/probe-rs/probe-rs/blob/644ad1a42c603a2475f9a89a0a3a71802db52749/probe-rs/src/probe/stlink/mod.rs#L71-L75

https://github.com/probe-rs/probe-rs/blob/644ad1a42c603a2475f9a89a0a3a71802db52749/probe-rs/src/probe/daplink/mod.rs#L229-L257

We should move this somewhere else, so that the initialisation of the debug interface is done in the same way, regardless of the probe. My suggestion would be to move it to the `ArmCommunicationInterface` struct.

Reply:
Alright, good idea! Seems like we do not always do the same and could well be that this is the issue with the chip not being able to halt on JLink flashing!

The question is: which of the different bits exactly do we need?